File: core/agent-core-ts/ts/skills.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { parse as parseYaml } from "yaml";
import type { Skill, SkillRegistryLike } from "./types.js";

const LOCAL_SKILL_DIR_CANDIDATES = ["skills", ".agents/skills", ".claude/skills", ".codex/skills"] as const;
const HOME_SKILL_DIR_CANDIDATES = [".codex/skills", ".agents/skills", ".claude/skills"] as const;

function asAbsolute(inputPath: string, cwd = process.cwd()): string {
  if (!inputPath.trim()) return "";
  return path.isAbsolute(inputPath) ? inputPath : path.resolve(cwd, inputPath);
}

function existsDirectory(dirPath: string): boolean {
  try {
    return fs.statSync(dirPath).isDirectory();
  } catch {
    return false;
  }
}

function collectAncestorDirectories(startDir: string): string[] {
  const result: string[] = [];
  let current = path.resolve(startDir);
  while (true) {
    result.push(current);
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return result;
}

function pushUnique(result: string[], value: string): void {
  if (!value) return;
  if (!result.includes(value)) {
    result.push(value);
  }
}

function parseSkillsDirsFromEnv(raw: string | undefined, cwd = process.cwd()): string[] {
  if (!raw) return [];
  return raw
    .split(path.delimiter)
    .map((item) => asAbsolute(item.trim(), cwd))
    .filter(Boolean);
}

function defaultSkillsDirs(startCwd = process.cwd()): string[] {
  const result: string[] = [];
  const cwd = path.resolve(startCwd);

  for (const envDir of parseSkillsDirsFromEnv(process.env.AGENT_SKILLS_DIR, cwd)) {
    pushUnique(result, envDir);
  }

  for (const baseDir of collectAncestorDirectories(cwd)) {
    for (const relativeDir of LOCAL_SKILL_DIR_CANDIDATES) {
      pushUnique(result, path.join(baseDir, relativeDir));
    }
  }

  const codexHome = process.env.CODEX_HOME?.trim();
  if (codexHome) {
    pushUnique(result, path.join(asAbsolute(codexHome, cwd), "skills"));
  }

  const homeDir = os.homedir();
  for (const relativeDir of HOME_SKILL_DIR_CANDIDATES) {
    pushUnique(result, path.join(homeDir, relativeDir));
  }

  return result.filter((dirPath) => existsDirectory(dirPath));
}

function enabledSkillNamesFromEnv(): string[] {
  const raw = process.env.AGENT_ENABLED_SKILLS ?? "";
  return raw
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean);
}

function parseSkillMarkdown(skillPath: string): Skill | null {
  const raw = fs.readFileSync(skillPath, "utf8").trim();
  let metadata: Record<string, unknown> = {};
  let body = raw;

  if (raw.startsWith("---")) {
    const parts = raw.split("---", 3);
    if (parts.length === 3) {
      metadata = (parseYaml(parts[1]?.trim() ?? "") as Record<string, unknown> | null) ?? {};
      body = parts[2] ?? "";
    }
  }

  const name = String(metadata.name ?? path.basename(path.dirname(skillPath))).trim();
  if (!name) return null;

  return {
    name,
    description: String(metadata.description ?? "").trim(),
    content: body.trim(),
    path: skillPath,
    metadata
  };
}

function walkSkillFiles(dir: string, result: string[] = []): string[] {
  if (!fs.existsSync(dir)) return result;

  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walkSkillFiles(fullPath, result);
      continue;
    }
    if (entry.isFile() && entry.name === "SKILL.md") {
      result.push(fullPath);
    }
  }

  return result;
}

export class SkillRegistry implements SkillRegistryLike {
  private readonly skillsDirs: string[];

  constructor(skillsDirOrDirs?: string | string[]) {
    if (Array.isArray(skillsDirOrDirs)) {
      this.skillsDirs = skillsDirOrDirs.map((item) => asAbsolute(item)).filter((item) => existsDirectory(item));
      return;
    }

    if (typeof skillsDirOrDirs === "string" && skillsDirOrDirs.trim()) {
      this.skillsDirs = [asAbsolute(skillsDirOrDirs)].filter((item) => existsDirectory(item));
      return;
    }

    this.skillsDirs = defaultSkillsDirs();
  }

  getResolvedSkillDirs(): string[] {
    return [...this.skillsDirs];
  }

  listSkills(options: { enabledOnly?: boolean; enabledNames?: string[] } = {}): Skill[] {
    const enabledSet = new Set((options.enabledNames ?? enabledSkillNamesFromEnv()).map((item) => item.trim()).filter(Boolean));
    const byName = new Map<string, Skill>();
    for (const skillDir of this.skillsDirs) {
      for (const skillPath of walkSkillFiles(skillDir)) {
        const skill = parseSkillMarkdown(skillPath);
        if (!skill) continue;
        if (!byName.has(skill.name)) {
          byName.set(skill.name, skill);
        }
      }
    }
    const skills = [...byName.values()].sort((a, b) => a.name.localeCompare(b.name));

    if (enabledSet.size > 0) {
      return skills.filter((skill) => enabledSet.has(skill.name));
    }

    if (options.enabledOnly) {
      return [];
    }

    return skills;
  }

  getSkill(name: string): Skill | null {
    const target = name.trim();
    return this.listSkills().find((skill) => skill.name === target) ?? null;
  }

  renderPromptContext(options: { enabledNames?: string[] } = {}): string {
    const skills = this.listSkills({
      enabledOnly: Boolean(options.enabledNames?.length || enabledSkillNamesFromEnv().length),
      enabledNames: options.enabledNames
    });
    if (skills.length === 0) return "";

    const lines = ["Available skills:"];
    for (const skill of skills) {
      lines.push(skill.description ? `- ${skill.name}: ${skill.description}` : `- ${skill.name}`);
    }
    lines.push("Use the list_skills or read_skill tools when you need exact skill instructions.");
    return lines.join("\n");
  }
}

File: core/agent-core-python/agent_core/skills.py
from __future__ import annotations

import os
from pathlib import Path
from typing import Any, Sequence

import yaml

from .types import Skill


LOCAL_SKILL_DIR_CANDIDATES: tuple[str, ...] = ("skills", ".agents/skills", ".claude/skills", ".codex/skills")
HOME_SKILL_DIR_CANDIDATES: tuple[str, ...] = (".codex/skills", ".agents/skills", ".claude/skills")


def _is_dir(path: Path) -> bool:
    try:
        return path.is_dir()
    except OSError:
        return False


def _parse_env_skill_dirs(raw: str | None, cwd: Path) -> list[Path]:
    if not raw:
        return []
    result: list[Path] = []
    for item in raw.split(os.pathsep):
        value = item.strip()
        if not value:
            continue
        path = Path(value)
        if not path.is_absolute():
            path = (cwd / path).resolve()
        result.append(path)
    return result


def _default_skills_dirs(start_cwd: Path | None = None) -> list[Path]:
    cwd = (start_cwd or Path.cwd()).resolve()
    result: list[Path] = []
    seen: set[Path] = set()

    def _push(candidate: Path) -> None:
        resolved = candidate.resolve()
        if resolved in seen:
            return
        seen.add(resolved)
        if _is_dir(resolved):
            result.append(resolved)

    for env_dir in _parse_env_skill_dirs(os.getenv("AGENT_SKILLS_DIR"), cwd):
        _push(env_dir)

    for base in [cwd, *cwd.parents]:
        for relative_dir in LOCAL_SKILL_DIR_CANDIDATES:
            _push(base / relative_dir)

    codex_home = os.getenv("CODEX_HOME")
    if codex_home:
        _push((Path(codex_home).expanduser() / "skills").resolve())

    home = Path.home()
    for relative_dir in HOME_SKILL_DIR_CANDIDATES:
        _push(home / relative_dir)

    return result


def _enabled_skill_names_from_env() -> list[str]:
    raw = os.getenv("AGENT_ENABLED_SKILLS", "")
    return [item.strip() for item in raw.split(",") if item.strip()]


def _parse_skill_markdown(path: Path) -> Skill | None:
    raw = path.read_text(encoding="utf-8").strip()
    metadata: dict[str, Any] = {}
    body = raw

    if raw.startswith("---"):
        parts = raw.split("---", 2)
        if len(parts) == 3:
            _, frontmatter, body = parts
            parsed = yaml.safe_load(frontmatter.strip()) or {}
            if isinstance(parsed, dict):
                metadata = {str(key): value for key, value in parsed.items()}

    name = str(metadata.get("name") or path.parent.name).strip()
    description = str(metadata.get("description") or "").strip()
    if not name:
        return None

    return Skill(
        name=name,
        description=description,
        content=body.strip(),
        path=path,
        metadata=metadata,
    )


class SkillRegistry:
    def __init__(self, skills_dir: str | Path | Sequence[str | Path] | None = None) -> None:
        if isinstance(skills_dir, Sequence) and not isinstance(skills_dir, (str, Path)):
            self._skills_dirs = [Path(item).expanduser().resolve() for item in skills_dir if _is_dir(Path(item).expanduser().resolve())]
            return

        if isinstance(skills_dir, (str, Path)) and str(skills_dir).strip():
            value = Path(skills_dir).expanduser()
            self._skills_dirs = [value.resolve()] if _is_dir(value.resolve()) else []
            return

        self._skills_dirs = _default_skills_dirs()

    def resolved_skill_dirs(self) -> list[Path]:
        return list(self._skills_dirs)

    def list_skills(
        self,
        *,
        enabled_only: bool = False,
        enabled_names: Sequence[str] | None = None,
    ) -> list[Skill]:
        enabled_set = {item.strip() for item in enabled_names or _enabled_skill_names_from_env() if item.strip()}
        if not self._skills_dirs:
            return []

        skills_by_name: dict[str, Skill] = {}
        for skill_dir in self._skills_dirs:
            for path in sorted(skill_dir.rglob("SKILL.md")):
                skill = _parse_skill_markdown(path)
                if skill is None or skill.name in skills_by_name:
                    continue
                skills_by_name[skill.name] = skill

        skills = sorted(skills_by_name.values(), key=lambda item: item.name)

        if enabled_set:
            skills = [skill for skill in skills if skill.name in enabled_set]
        return skills

    def get_skill(self, name: str) -> Skill | None:
        target = name.strip()
        for skill in self.list_skills():
            if skill.name == target:
                return skill
        return None

    def render_prompt_context(self, *, enabled_names: Sequence[str] | None = None) -> str:
        skills = self.list_skills(
            enabled_only=bool(enabled_names or _enabled_skill_names_from_env()),
            enabled_names=enabled_names,
        )
        if not skills:
            return ""

        lines = ["Available skills:"]
        for skill in skills:
            summary = f"- {skill.name}"
            if skill.description:
                summary += f": {skill.description}"
            lines.append(summary)
        lines.append("Use the list_skills or read_skill tools when you need exact skill instructions.")
        return "\n".join(lines)
